RearrangeCarGroup8Movements.py

```
@@ -0,0 +1,206 @@
+        space position
+
+        Otherwise picks up the random not arranged car and places it to the empty space
+        in this case the number of not arranged cars stays the same
+
+        Complexity is O(1)
+        '''
+
+        if self.final_car_order[self.empty_position] == 0:
+            car_to_move_position = self._get_not_arranged_car_index()
+            car_to_move = self.current_car_order[car_to_move_position]
+        else:
+            car_to_move = self.final_car_order[self.empty_position]
+            car_to_move_position = self.car_position_dict[car_to_move]
+            self.not_arranged_cars.difference_update([car_to_move]) # O(1)
+
+        self._move_car_to_empty_slot(car_to_move_position)
+        return Movement(car_to_move, car_to_move_position, self.empty_position);
+
+    def _variables_to_init_state(self):
+
+        '''Resets variables to the same state when class was initialized
+        Needed for running rearrange_all_cars function again
+        '''
+
+        self.current_car_order = self.initial_car_order
+        self.empty_position = self.current_car_order.index(0)
+        self.car_position_dict = {car:pos for pos, car in \
+                                  enumerate(self.initial_car_order) if car} # time O(n)
+        self.not_arranged_cars = {car for i, car in enumerate(self.initial_car_order)\
+                                  if self.initial_car_order[i] != self.final_car_order[i]\
+                                  and car} # time O(n)
+
+    def rearrange_all_cars(self):
+
+        '''Rearranges all cars and prints the moves
+        Note 1: In case there are more than 1000 moves to print
+        logging into the file is done
+        Note 2: Possible to run the function many times
+
+        '''
+        
+        if self.initial_car_order == self.final_car_order:
+            assert ValueError, 'Cars are already in the final order, no moves are required'

```

Already arranged cars is a desired situation, I don't think we should raise an error (I suppose this is raising, as the syntax feels unfamiliar to me - would be great if anyone explained)

---

Good point, the assert is not needed here probably, I fix that)

---

+1 for removing the assert. I think we can get rid of the checking entirely :)